syntax="proto3";
option go_package = "./pb";
package pb;
import "freeconf/pb/meta.proto";
import "freeconf/pb/common.proto";
import "freeconf/pb/fs.proto";

/*
  FreeCONF full developer API made available thru gRPC.  Services are roughly broken
  up by package.
*/

////////////
// Parser

service Parser {
      rpc LoadModule (LoadModuleRequest) returns (LoadModuleResponse) {}
}

message LoadModuleRequest {
      string dir = 1;
      string name = 2;
}

message LoadModuleResponse {
      Module module = 1;
      uint64 moduleHnd = 2;
}

////////////
// Handles
service Handles {
      rpc Release (ReleaseRequest) returns (ReleaseResponse) {}
}

message ReleaseResponse {
}

message ReleaseRequest {
      uint64 hnd = 1;
}

////////////


service Node {
      rpc NewBrowser(NewBrowserRequest) returns (NewBrowserResponse) {}
      rpc BrowserRoot(BrowserRootRequest) returns (BrowserRootResponse) {}
      rpc SelectionEdit(SelectionEditRequest) returns (SelectionEditResponse) {}
      rpc Action(ActionRequest) returns (ActionResponse) {}
      rpc Notification(NotificationRequest) returns (stream NotificationResponse) {}
      rpc Find(FindRequest) returns (FindResponse) {}
      rpc NewNode(NewNodeRequest) returns (NewNodeResponse) {}
      rpc GetBrowser(GetBrowserRequest) returns (GetBrowserResponse) {}
      rpc GetModule(GetModuleRequest) returns (GetModuleResponse) {}
      rpc GetSelection(GetSelectionRequest) returns (GetSelectionResponse) {}
      rpc ReleaseSelection(ReleaseSelectionRequest) returns (ReleaseSelectionResponse) {}
}

enum SelectionEditOp {
      UPSERT_INTO = 0;
      UPSERT_FROM = 1;
      INSERT_INTO = 2;
      INSERT_FROM = 3;
      UPSERT_INTO_SET_DEFAULTS = 4;
      UPSERT_FROM_SET_DEFAULTS = 5;
      UPDATE_INTO = 6;
      UPDATE_FROM = 7;
      REPLACE_FROM = 8;
}

message NewBrowserRequest {
      uint64 moduleHnd = 1;
      uint64 nodeHnd = 3;
}

message NewBrowserResponse {
      uint64 browserHnd = 1;
}

message ReleaseSelectionRequest {
      uint64 selHnd = 1;
}

message ReleaseSelectionResponse {      
}

message SelectionEditRequest {
      SelectionEditOp op = 1;
      uint64 selHnd = 2;
      uint64 nodeHnd = 3;
}

message SelectionEditResponse {
}

message BrowserRootRequest {
      uint64 browserHnd = 1;
}

message BrowserRootResponse {
      uint64 selHnd = 1;
}

message NewNodeRequest {      
}

message NewNodeResponse {
      uint64 nodeHnd = 1;
}

message FindRequest {
      uint64 selHnd = 1;
      string path = 2;
}

message FindResponse {
      uint64 selHnd = 1;
}

message ActionRequest {
      uint64 selHnd = 1;
      uint64 inputNodeHnd = 2;      
}

message ActionResponse {
      uint64 outputSelHnd = 1;
}

message GetBrowserRequest {
      uint64 browserHnd = 1;
}

message GetBrowserResponse {
      uint64 moduleHnd = 1;
}

message GetModuleRequest {
      uint64 moduleHnd = 1;
}

message GetModuleResponse {
      Module module = 1;
}

message GetSelectionRequest {
      uint64 selHnd = 1;
}

message GetSelectionResponse {      
      uint64 nodeHnd = 1;
      bool remoteNode = 2; // true is node originaly from Go, false if from X lang
      uint64 browserHnd = 3;      
      bool insideList = 4;
      Path path = 5;
}

message NotificationRequest {
      uint64 selHnd = 1;
}

message NotificationResponse {
      uint64 selHnd = 1;
}

////////////
service Device {
      rpc NewDevice(NewDeviceRequest) returns (NewDeviceResponse) {}
      rpc DeviceAddBrowser(DeviceAddBrowserRequest) returns (DeviceAddBrowserResponse) {}
      rpc DeviceGetBrowser(DeviceGetBrowserRequest) returns (DeviceGetBrowserResponse) {}
      rpc ApplyStartupConfig(ApplyStartupConfigRequest) returns (ApplyStartupConfigResponse) {}
}

message NewDeviceRequest {
      string yangPath = 1;
}

message NewDeviceResponse {
      uint64 deviceHnd = 1;
}

message DeviceAddBrowserRequest {
      uint64 deviceHnd = 1;
      uint64 browserHnd = 2;
}

message DeviceAddBrowserResponse {      
}

message DeviceGetBrowserRequest {
      uint64 deviceHnd = 1;
      string moduleIdent = 2;
}

message DeviceGetBrowserResponse {
      uint64 browserHnd = 1;
}

message ApplyStartupConfigRequest {
      uint64 deviceHnd = 1;
      string configFile = 2;
}

message ApplyStartupConfigResponse {
}

////////////
service NodeUtil {
      rpc JSONRdr(JSONRdrRequest) returns (JSONRdrResponse) {}
      rpc JSONWtr(JSONWtrRequest) returns (JSONWtrResponse) {}
}

message JSONRdrRequest {
      FileHandle file = 1;
}

message JSONRdrResponse {
      uint64 nodeHnd = 1;
}


message JSONWtrRequest {
      string fname = 1;
}

message JSONWtrResponse {
      uint64 nodeHnd = 1;
}

////////////
service Restconf {
      rpc NewServer(NewServerRequest) returns (NewServerResponse) {}
}

message NewServerRequest {
      uint64 deviceHnd = 1;
}

message NewServerResponse {
      uint64 serverHnd = 1;
}

